### PR TITLE
Fix futures_io::AsyncSeek implementaion for Compat

### PR DESCRIPTION
**Pull request by dhruv9vats**
## Motivation
Aims to address #5764.

## Solution
Waiting for a seek operation to complete by calling `poll_complete` before calling `start_seek`.

### CODE REVIEW

File: tokio-util/tests/compat.rs
Comment (dhruv9vats) [inline tokio-util/tests/compat.rs:1]: This is not intended as a test, but is merely an attempt to reproduce the example provided in the issue.

However, I did not see this test fail, even without the fix above. 

Thus, ended up using the (very helpful) repo that was provided in the issue: https://github.com/tmpfs/tokio-compat-seek (which does reliably reproduce the error), pointing the `Cargo.toml` in there to use the local tokio repo.

Are there any existing tests I could take inspiration from for this?

File: tokio-util/src/compat.rs [hunk 1]
```diff
@@ -226,13 +226,18 @@ impl<T: tokio::io::AsyncSeek> futures_io::AsyncSeek for Compat<T> {
         cx: &mut Context<'_>,
         pos: io::SeekFrom,
     ) -> Poll<io::Result<u64>> {
-        if self.seek_pos != Some(pos) {
-            self.as_mut().project().inner.start_seek(pos)?;
-            *self.as_mut().project().seek_pos = Some(pos);
+        match self.as_mut().project().inner.poll_complete(cx) {
```
[the remark this answers is not in the document]
Comment (Darksonn) [reply tokio-util/src/compat.rs:229]: Prefer the `ready!` macro.

File: tokio-util/Cargo.toml [hunk 2]
```diff
@@ -56,6 +58,7 @@ async-stream = "0.3.0"
 futures = "0.3.0"
 futures-test = "0.3.5"
 parking_lot = "0.12.0"
+tempfile = "3.1.0"
```
Comment (dhruv9vats) [inline tokio-util/Cargo.toml:61]: Adding a dependency to add a test feels overkill. 
But the thing is, the test only fails for a `tokio::fs::file::File`.
Using something like `tokio::io::util::buf_writer::BufWriter` or `std::io::cursor::Cursor` seem to be working fine on latest `upstream/master`:
```rust
#[tokio::test]
async fn compat_seek() -> futures_util::io::Result<()> {
    let mut cursor = Cursor::new(Vec::new()).compat_write();

    cursor.write_all(&[0, 1, 2, 3, 4, 5]).await?;
    cursor.write_all(&[6, 7]).await?;

    assert_eq!(cursor.stream_position().await?, 8);

    // Modify elements at position 2.
    assert_eq!(cursor.seek(SeekFrom::Start(2)).await.unwrap(), 2);
    cursor.write_all(&[8, 9]).await.unwrap();

    cursor.flush().await.unwrap();

    // Verify we still have 8 elements
    assert_eq!(cursor.seek(SeekFrom::End(0)).await?, 8);
    // Seek back to the start of the file to read and verify contents.
    cursor.seek(SeekFrom::Start(0)).await?;

    let mut buf = Vec::new();
    let num_bytes = cursor.read_to_end(&mut buf).await?;
    assert_eq!(&buf[..num_bytes], &[0, 1, 8, 9, 4, 5, 6, 7]);

    Ok(())
}
```

Comment (dhruv9vats) [reply tokio-util/Cargo.toml:61]: @Darksonn, is there some alternative way to test this? And would like more sense to skip adding one if there is none?

File: tokio-util/tests/compat.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
+#![cfg(all(feature = "compat", not(tokio_wasi)))] // WASI does not support all fs operations
```
Comment (Darksonn) [inline tokio-util/tests/compat.rs:1]: The `tokio_wasi` marker only works in the main Tokio crate. (via the build script)

File: tokio-util/Cargo.toml [hunk 4]
```diff
@@ -42,7 +42,9 @@ futures-io = { version = "0.3.0", optional = true }
 futures-util = { version = "0.3.0", optional = true }
 pin-project-lite = "0.2.7"
 slab = { version = "0.4.4", optional = true } # Backs `DelayQueue`
-tracing = { version = "0.1.25", default-features = false, features = ["std"], optional = true }
+tracing = { version = "0.1.25", default-features = false, features = [
+    "std",
+], optional = true }
```
Comment (Darksonn) [inline tokio-util/Cargo.toml:47]: I'd rather keep this one unchanged.

File: tokio-util/Cargo.toml (hunk 2 above)
Comment (Darksonn) [reply tokio-util/Cargo.toml:61]: This dependency already exists in the main Tokio crate, so adding it here doesn't _actually_ add a new dependency.

**Review by Darksonn — APPROVED** [review]
Thanks.

**Pull request merged**